r/src/array_from_vector.cpp

```diff
@@ -41,75 +49,190 @@ inline bool is_na<int>(int value) {
 
-    // catch up
-    for (R_xlen_t j = 0; j < i; j++, null_bitmap_writer.Next()) {
-      null_bitmap_writer.Set();
+    auto* int_builder = checked_cast<Int64Builder*>(builder);
+
+    R_xlen_t n = XLENGTH(x);
+    const auto* data = (REAL(x));
+
+    RETURN_NOT_OK(int_builder->Reserve(n));
+    for (R_xlen_t i = 0; i < n; i++) {
+      const auto value = arrow::util::SafeCopy<int64_t>(data[i]);
+      if (value == NA_INT64) {
+        int_builder->UnsafeAppendNull();
+      } else {
+        int_builder->UnsafeAppend(value);
+      }
     }
 
-    // resume offset filling
-    for (; i < n; i++, ++p_offset, null_bitmap_writer.Next()) {
-      SEXP s = STRING_ELT(vec, i);
-      if (s == NA_STRING) {
-        null_bitmap_writer.Clear();
-        *p_offset = current_offset;
-        null_count++;
+    return Status::OK();
+  }
+
+  Status Visit(const arrow::DoubleType& type) {
+    ARROW_RETURN_IF(TYPEOF(x) != REALSXP,
+                    Status::Invalid("Expected vector of type REALSXP"));
+
+    auto* double_builder = checked_cast<DoubleBuilder*>(builder);
+
+    R_xlen_t n = XLENGTH(x);
+    const auto* data = (REAL(x));
+
+    RETURN_NOT_OK(double_builder->Reserve(n));
+    for (R_xlen_t i = 0; i < n; i++) {
+      const auto value = data[i];
+      if (ISNA(value)) {
+        double_builder->UnsafeAppendNull();
       } else {
-        null_bitmap_writer.Set();
-        *p_offset = current_offset += LENGTH(s);
+        double_builder->UnsafeAppend(value);
       }
     }
 
-    null_bitmap_writer.Finish();
+    return Status::OK();
+  }
+
+  template <typename T>
+  arrow::enable_if_base_binary<T, Status> Visit(const T& type) {
+    using BuilderType = typename TypeTraits<T>::BuilderType;
+    auto* binary_builder = checked_cast<BuilderType*>(builder);
+
+    R_xlen_t n = XLENGTH(x);
+    builder->Reserve(n);
+    for (R_xlen_t i = 0; i < n; i++) {
+      SEXP s = STRING_ELT(x, i);
+      if (s == NA_STRING) {
+        RETURN_NOT_OK(binary_builder->AppendNull());
+        continue;
+      }
+
+      RETURN_NOT_OK(binary_builder->Append(CHAR(s), LENGTH(s)));
+    }
+
+    return Status::OK();
   }
 
-  // ----- data buffer
-  if (current_offset > 0) {
-    STOP_IF_NOT_OK(AllocateBuffer(current_offset, &value_buffer));
-    p_offset = reinterpret_cast<int32_t*>(offset_buffer->mutable_data());
-    auto p_data = reinterpret_cast<char*>(value_buffer->mutable_data());
+  template <typename T>
+  arrow::enable_if_base_list<T, Status> Visit(const T& type) {
+    using BuilderType = typename TypeTraits<T>::BuilderType;
+
+    auto* list_builder = checked_cast<BuilderType*>(builder);
+    auto* value_builder = list_builder->value_builder();
+    auto value_type = type.value_type();
+
+    R_xlen_t n = XLENGTH(x);
+
+    ARROW_RETURN_IF(TYPEOF(x) != VECSXP,
+                    Status::Invalid("Expected vector of type VECSXP"));
+    ARROW_RETURN_IF(n == 0, Status::Invalid("0-length vector input not supported"));
```

Comment:
Why not?

Reply:
Removed, the compat check required this. Now this is only required in `InferArrowType` and it supports empty list with an explicit type hint.